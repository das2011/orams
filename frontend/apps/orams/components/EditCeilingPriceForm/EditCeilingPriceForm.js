import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Form, Control } from 'react-redux-form'
import ErrorBox from 'shared/form/ErrorBox'
import Textfield from 'shared/form/Textfield'
import formProps from 'shared/form/formPropsSelector'
import { validMoreThanMinPrice } from 'shared/validators'
import styles from './EditCeilingPriceForm.scss'

class EditCeilingPriceForm extends Component {
  renderHeading() {
    const { serviceToEdit, supplier, goToStep, priceData } = this.props
    const { region } = priceData
    const supplierName = supplier ? supplier.name : ''

    return (
      <header>
        <h1 className="au-display-xl">
          {`${supplierName} - ${region.state} ${region.name} - ${serviceToEdit.serviceName}`}
        </h1>
        <h2 className="au-display-lg"> Update ceiling prices - edit ceiling price</h2>
        <div className={styles.stepTitle}>Step 3 of 3</div>
        <div className={styles.backLink} onClick={() => goToStep(2)}>
          Back to pricing information
        </div>
      </header>
    )
  }

  render() {
    const { model, action, submitClicked, priceData, handlePriceSubmit } = this.props
    let hasFocused = false

    const setFocus = e => {
      if (!hasFocused) {
        hasFocused = true
        e.focus()
      }
    }

    return (
      <div className={styles.container}>
        {this.renderHeading()}
        <article role="main">
          <ErrorBox model={model} setFocus={setFocus} submitClicked={submitClicked} />
          <Form
            model={model}
            action={action}
            id="EditCeilingPrice__create"
            onSubmit={data => handlePriceSubmit(data, priceData.capPriceId)}
          >
            <Textfield
              model={`${model}.ceilingPrice`}
              name="ceilingPrice"
              id="ceilingPrice"
              htmlFor="ceilingPrice"
              label="Enter a new ceiling price including GST"
              description={`The new ceiling price must not be less than $${priceData.price}`}
              defaultValue={priceData.capPrice}
              validators={{ validMoreThanMinPrice: validMoreThanMinPrice(priceData.price) }}
              messages={{
                validMoreThanMinPrice: `Ceiling price must be valid and more than $${priceData.price}`
              }}
            />
            <div className="question-heading au-text-input__label">
              Check box if you want to set current price to match
            </div>
            <div className={styles.ceilingPriceCheckbox}>
              <label className="au-control-input au-control-input--full" htmlFor="setCurrentPriceToCeiling">
                <Control.checkbox
                  model={`${model}.setCurrentPriceToCeiling`}
                  id="setCurrentPriceToCeiling"
                  name="setCurrentPriceToCeiling"
                  value="yes"
                  mapProps={{
                    className: 'au-control-input__input match-ceiling-price__checkbox'
                  }}
                />
                <span className="au-control-input__text">Set current price to match</span>
              </label>
            </div>
            <div className="row">
              <div className="col-xs-12 col-sm-12 col-md-11">
                <button type="submit" className="au-btn">
                  Save and return to region list
                </button>
              </div>
            </div>
          </Form>
        </article>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  ...formProps(state, 'editCeilingPriceForm')
})

export { mapStateToProps, EditCeilingPriceForm as Form }

export default connect(mapStateToProps)(EditCeilingPriceForm)
