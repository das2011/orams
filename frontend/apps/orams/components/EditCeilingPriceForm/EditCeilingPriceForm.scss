.container {
  margin-bottom: 100px;
}

.stepTitle {
  font-size: 20px;
  color: #777777;
}

.backLink {
  font-size: 17px;
  font-weight: bold;
  margin-top: 15px;
  margin-bottom: 49px;
  cursor: pointer;
  text-decoration: underline;
}

.button {
  font-size: 20px;
  line-height: 0.8;
  background-color: #f5d815;
  color: #000000;
  padding: 12px 17px;
  width: 100%;
  text-align: center;
  margin-right: 21px;
  border: none;
}

@media screen and (min-width: 768px) {
  .button {
      width: 44%;
  }
}
