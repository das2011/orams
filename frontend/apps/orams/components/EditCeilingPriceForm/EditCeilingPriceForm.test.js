import React from 'react'
import Enzyme, { mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import expect from 'expect'
import { Provider } from 'react-redux'
import snapshotRenderer from 'react-test-renderer'
import configureStore from 'orams/store/store.prod.js' // dont want the debugger
import EditCeilingPriceForm from './EditCeilingPriceForm'

Enzyme.configure({ adapter: new Adapter() })

describe('#EditCeilingPriceForm', () => {
  const gotoStepSpy = jest.fn()
  const handlePriceSubmitSpy = jest.fn()
  const props = {
    supplier: {
      name: 'Supplier Name'
    },
    priceData: {
      capPrice: '1000',
      capPriceId: 1,
      price: '500',
      region: {
        name: 'Metro',
        state: 'NSW'
      }
    },
    serviceToEdit: {
      serviceName: 'Service To Edit'
    },
    goToStep: gotoStepSpy
  }

  const store = configureStore()

  test('renders with pricing details', () => {
    const tree = snapshotRenderer
      .create(
        <Provider store={store}>
          <EditCeilingPriceForm {...props} handlePriceSubmit={handlePriceSubmitSpy} />
        </Provider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test('goToStep() is called with correct parameters', () => {
    const wrapper = mount(
      <Provider store={store}>
        <EditCeilingPriceForm {...props} handlePriceSubmit={handlePriceSubmitSpy} />
      </Provider>
    )

    wrapper.find('.backLink').simulate('click')

    expect(gotoStepSpy).toBeCalledWith(2)
  })

  test('error validation when new cap price is less than current price', () => {
    const wrapper = mount(
      <Provider store={store}>
        <EditCeilingPriceForm {...props} handlePriceSubmit={handlePriceSubmitSpy} />
      </Provider>
    )

    const ceilingPriceInput = wrapper.find('input[type="text"]')
    ceilingPriceInput.simulate('change', { target: { value: '100' } })
    ceilingPriceInput.simulate('blur')

    const validationMessageElement = wrapper.find('.validation-message')

    expect(validationMessageElement.length).toBe(1)
  })

  describe('handlePriceSubmit()', () => {
    test('called with correct parameters when set current price to match is checked', () => {
      const checkedHandlePriceSubmitSpy = jest.fn()
      const expectedPriceData = props.priceData
      const wrapper = mount(
        <Provider store={store}>
          <EditCeilingPriceForm {...props} handlePriceSubmit={checkedHandlePriceSubmitSpy} />
        </Provider>
      )

      wrapper.find('input.match-ceiling-price__checkbox').simulate('change', { target: { checked: true } })
      wrapper.find('.au-btn').simulate('submit')

      expect(checkedHandlePriceSubmitSpy).toBeCalledWith(
        {
          ceilingPrice: expectedPriceData.capPrice,
          setCurrentPriceToCeiling: true
        },
        expectedPriceData.capPriceId
      )
    })

    test('called with correct parameters when set current price to match is not checked', () => {
      const uncheckedHandlePriceSubmitSpy = jest.fn()
      const expectedPriceData = props.priceData
      const wrapper = mount(
        <Provider store={store}>
          <EditCeilingPriceForm {...props} handlePriceSubmit={uncheckedHandlePriceSubmitSpy} />
        </Provider>
      )

      wrapper.find('input.match-ceiling-price__checkbox').simulate('change', { target: { checked: false } })
      wrapper.find('.au-btn').simulate('submit')

      expect(uncheckedHandlePriceSubmitSpy).toBeCalledWith(
        {
          ceilingPrice: expectedPriceData.capPrice,
          setCurrentPriceToCeiling: false
        },
        expectedPriceData.capPriceId
      )
    })
  })
})
