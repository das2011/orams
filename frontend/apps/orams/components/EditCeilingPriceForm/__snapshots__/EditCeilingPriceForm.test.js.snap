// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#EditCeilingPriceForm renders with pricing details 1`] = `
<div
  className="container"
>
  <header>
    <h1
      className="au-display-xl"
    >
      Supplier Name - NSW Metro - Service To Edit
    </h1>
    <h2
      className="au-display-lg"
    >
       Update ceiling prices - edit ceiling price
    </h2>
    <div
      className="stepTitle"
    >
      Step 3 of 3
    </div>
    <div
      className="backLink"
      onClick={[Function]}
    >
      Back to pricing information
    </div>
  </header>
  <article
    role="main"
  >
    <form
      action={undefined}
      id="EditCeilingPrice__create"
      noValidate={undefined}
      onReset={[Function]}
      onSubmit={[Function]}
    >
      <div
        className="field"
      >
        <label
          className="question-heading au-text-input__label"
          htmlFor="ceilingPrice"
          id="ceilingPrice-label"
        >
          Enter a new ceiling price including GST
        </label>
        <div
          className="au-text-input__hint"
          id="ceilingPrice-hint"
        >
          The new ceiling price must not be less than $500
        </div>
        <input
          aria-describedby="ceilingPrice-label"
          className="au-text-input au-text-input--block"
          disabled={false}
          id="ceilingPrice"
          maxLength={undefined}
          name="ceilingPrice"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          pattern={null}
          readOnly={false}
          type="text"
          value="1000"
        />
      </div>
      <div
        className="question-heading au-text-input__label"
      >
        Check box if you want to set current price to match
      </div>
      <div
        className="ceilingPriceCheckbox"
      >
        <label
          className="au-control-input au-control-input--full"
          htmlFor="setCurrentPriceToCeiling"
        >
          <input
            checked={false}
            className="au-control-input__input match-ceiling-price__checkbox"
            disabled={false}
            id="setCurrentPriceToCeiling"
            name="setCurrentPriceToCeiling"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyPress={[Function]}
            type="checkbox"
            value="yes"
          />
          <span
            className="au-control-input__text"
          >
            Set current price to match
          </span>
        </label>
      </div>
      <div
        className="row"
      >
        <div
          className="col-xs-12 col-sm-12 col-md-11"
        >
          <button
            className="au-btn"
            type="submit"
          >
            Save and return to region list
          </button>
        </div>
      </div>
    </form>
  </article>
</div>
`;
